File: marketing/channels/linkedin.js
// =============================================================================
// Knol Marketing — LinkedIn Channel Adapter
// Posts articles/updates via LinkedIn API (free tier w/ OAuth app)
// Fallback: generates ready-to-post content for manual sharing
// =============================================================================

const https = require('https');

const CONFIG = {
  API_BASE: 'https://api.linkedin.com/v2',
  RATE_LIMIT: { perDay: 25 },
};

// Post a text update to LinkedIn personal profile
async function postUpdate(text, credentials) {
  const { accessToken, personUrn } = credentials;

  if (!accessToken || !personUrn) {
    return {
      success: false,
      error: 'Missing LinkedIn credentials — saving content for manual posting',
      content: text,
      manual: true,
    };
  }

  const body = JSON.stringify({
    author: personUrn,
    lifecycleState: 'PUBLISHED',
    specificContent: {
      'com.linkedin.ugc.ShareContent': {
        shareCommentary: { text },
        shareMediaCategory: 'NONE',
      },
    },
    visibility: {
      'com.linkedin.ugc.MemberNetworkVisibility': 'PUBLIC',
    },
  });

  return new Promise((resolve) => {
    const req = https.request({
      hostname: 'api.linkedin.com',
      path: '/v2/ugcPosts',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
        'X-Restli-Protocol-Version': '2.0.0',
        'Content-Length': Buffer.byteLength(body),
      },
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 201) {
          resolve({ success: true, postId: res.headers['x-restli-id'], statusCode: res.statusCode, headers: res.headers });
        } else {
          resolve({
            success: false,
            error: `HTTP ${res.statusCode}: ${data.substring(0, 200)}`,
            content: text,
            manual: true,
            rateLimited: res.statusCode === 429,
            statusCode: res.statusCode,
            headers: res.headers,
          });
        }
      });
    });

    req.on('error', (e) => resolve({ success: false, error: e.message, content: text, manual: true }));
    req.write(body);
    req.end();
  });
}

// Post an article with link preview
async function postArticle(title, text, url, credentials) {
  const fullText = `${text}\n\n${url}`;
  return postUpdate(fullText, credentials);
}

// Generate content formatted for LinkedIn (when no API access)
function formatForManualPost(content) {
  return {
    text: content,
    hashtags: '#opensource #memorylayer #rustlang #ai #llm #devtools',
    bestTimes: ['Tuesday 10am', 'Wednesday 12pm', 'Thursday 9am'],
  };
}

module.exports = { postUpdate, postArticle, formatForManualPost, CONFIG };

File: knol-enterprise/crates/service-marketing/src/channels/linkedin.rs
//! LinkedIn channel adapter — Bearer token → POST /v2/ugcPosts.

use tracing::{info, warn};

use super::{PublishContent, PublishResult};
use crate::config::ChannelCredentials;
use crate::error::MarketingError;

pub async fn publish(
    content: &PublishContent,
    http_client: &reqwest::Client,
    credentials: &ChannelCredentials,
) -> Result<PublishResult, MarketingError> {
    let token = credentials
        .linkedin_token
        .as_ref()
        .ok_or_else(|| MarketingError::Channel {
            channel: "linkedin".into(),
            message: "LinkedIn credentials not configured".into(),
        })?;
    let person_urn = credentials
        .linkedin_person_urn
        .as_deref()
        .unwrap_or("urn:li:person:unknown");

    let body = serde_json::json!({
        "author": person_urn,
        "lifecycleState": "PUBLISHED",
        "specificContent": {
            "com.linkedin.ugc.ShareContent": {
                "shareCommentary": { "text": content.text },
                "shareMediaCategory": "NONE"
            }
        },
        "visibility": {
            "com.linkedin.ugc.MemberNetworkVisibility": "PUBLIC"
        }
    });

    let resp = http_client
        .post("https://api.linkedin.com/v2/ugcPosts")
        .header("Authorization", format!("Bearer {}", token))
        .header("X-Restli-Protocol-Version", "2.0.0")
        .json(&body)
        .send()
        .await
        .map_err(|e| MarketingError::Channel {
            channel: "linkedin".into(),
            message: e.to_string(),
        })?;

    let status = resp.status();
    let post_id = resp
        .headers()
        .get("x-restli-id")
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string());

    if status.as_u16() == 201 {
        info!("LinkedIn: posted update");
        Ok(PublishResult::success("linkedin", post_id, None))
    } else {
        let text = resp.text().await.unwrap_or_default();
        warn!(
            "LinkedIn: HTTP {} — {}",
            status,
            &text[..text.len().min(200)]
        );
        Ok(PublishResult::failure(
            "linkedin",
            format!("HTTP {}: {}", status, &text[..text.len().min(200)]),
        ))
    }
}
